doc/cpp_simulation.rst

```diff
@@ -0,0 +1,139 @@
+          and current) and spike output.
+
+    **Types:**
+
+    .. cpp:type:: communicator_type = communication::communicator<communication::global_policy>
+
+        Type used for distributed communication of spikes and global synchronization.
+
+    .. cpp:type:: spike_export_function = std::function<void(const std::vector<spike>&)>
+
+        User-supplied callack function used as a sink for spikes generated
+        during a simulation. See :cpp:func:`set_local_spike_callback` and
+        :cpp:func:`set_global_spike_callback`.
+
+    **Constructor:**
+
+    .. cpp:function:: simulation(const recipe& rec, const domain_decomposition& decomp)
+
+    **Experimental inputs:**
+
+    .. cpp:function:: void inject_events(const pse_vector& events)
+
+        Add events directly to targets.
+        Must be called before calling :cpp:func:`simulation::run`, and must contain events that
+        are to be delivered at or after the current simulation time.
+
+    **Updating Model State:**
+
+    .. cpp:function:: void reset()
+
+        Reset the state of the simulation to its original state before
+        :cpp:func:`simulation::run` was called.
+
+    .. cpp:function:: time_type run(time_type tfinal, time_type dt)
+
+        Run the simulation from current simulation time to :cpp:var:`tfinal`,
+        with maximum time step size :cpp:var:`dt`.
+
+    .. cpp:function:: void set_binning_policy(binning_kind policy, time_type bin_interval)
+
+        Set event binning policy on all our groups.
+
+    **I/O:**
+
+    .. cpp:function:: sampler_association_handle add_sampler(\
+                        cell_member_predicate probe_ids,\
+                        schedule sched,\
+                        sampler_function f,\
+                        sampling_policy policy = sampling_policy::lax)
+
+        Note: sampler functions may be invoked from a different thread than that
+        which called :cpp:func:`simulation::run`.
+
+        (see the :ref:`sampling_api` documentation.)
+
+    .. cpp:function:: void remove_sampler(sampler_association_handle)
+
+        Remove a sampler.
+        (see the :ref:`sampling_api` documentation.)
+
+    .. cpp:function:: void remove_all_samplers()
+
+        Remove all samplers from probes.
+        (see the :ref:`sampling_api` documentation.)
+
+    .. cpp:function:: std::size_t num_spikes() const
+
+        The total number of spikes that occurred since either construction or
+        last resetting the model.
+
+    .. cpp:function:: void set_global_spike_callback(spike_export_function export_callback)
+
+        Register a callback that will perform an export of the global spike vector.
+        Will be called on the MPI rank/domain with id 0.
+
+    .. cpp:function:: void set_local_spike_callback(spike_export_function export_callback)
+
+        Register a callback that will perform an export of the rank local spike vector.
+        Will be called on each MPI rank/domain with a copy of the local spikes.
```

Comment:
What the local and global spike vectors are should be described (at least in brief), or have a link to a stub that promises to describe the semantics of distributed spike generation and exchange.